feat :  Implemented Logic for Filtering Requests by State and Username

AnujChhikara opened the pull request:
<!-- Date Here in dd mmm yyyy-->
Date: 27 Feb, 2025
<!--Developer Name Here-->
Developer Name: @AnujChhikara 

---

## Issue Ticket Number
<!--Issue ticket this PR closes-->
- #864 
## Description
This PR adds the logic to filter requests based on state and username. The UI for filtering was already in place, but the filtering functionality was not implemented. With this update:

- Users can now filter requests by state (e.g., pending, approved, rejected).
- Users can also filter requests by username, allowing them to quickly find relevant requests.
- Debouncing implemented: When a user types a username, we wait 300ms before sending the request to minimize API calls and improve performance.
<!--Description of the changes made in this PR-->

### Documentation Updated?

- [ ] Yes
- [x] No

<!--Additional notes about documentation update if applicable-->

### Under Feature Flag

- [x] Yes
- [ ] No

<!--Indicate if changes are under a feature flag-->

### Database Changes

- [ ] Yes
- [x] No

<!--Notes on any database changes-->

### Breaking Changes

- [ ] Yes
- [x] No

<!--Notes on breaking changes or related issue tickets if applicable-->

### Development Tested?

- [x] Yes
- [ ] No

<!--Confirmation of local testing during development-->

## Screenshots
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->
 - Filter by State

https://github.com/user-attachments/assets/c5153b17-e937-450e-b147-ee3d30af17a6



 - Filter by Username

https://github.com/user-attachments/assets/37c687de-12d9-490c-8056-7d7d0f15d0e1



 - Approving and rejecting Requests

https://github.com/user-attachments/assets/486b161e-ceae-4b3d-913a-604449ab1d75


</details>

<!--Attach or link to relevant screenshots or visual aids-->

## Test Coverage
<details>
<summary>Screenshot 1</summary>

<!-- Attach your screenshots here👇-->
![image](https://github.com/user-attachments/assets/fa9c41c2-dd75-4ffa-b789-b0243f98291d)

</details>

<!--Attach Details on test coverage and outcomes-->

## Additional Notes
- This is the last PR for this task 
<!--Any additional notes, considerations or explanations for reviewers-->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a new user suggestion feature based on the search term "randhir".
  - Improved request filtering to display only approved requests when the filter is applied.
  - Enhanced assignee search with a debounced input, providing timely and relevant user suggestions.

- **Style**
  - Refined the appearance of suggestion elements with adjusted spacing and font size for better readability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MayankBansal12 commented on requests/util.js:
```diff
@@ -43,6 +43,9 @@ function getQueryParamsString(requestType, query) {
   ) {
     queryParam += `&state=${query.state}`;
   }
+  if (query.requestedBy !== undefined && query.requestedBy !== null) {
+    queryParam += `&requestedBy=${query.requestedBy}`;
+  }
```
```suggestion
  if (query?.requestedBy != null) {
    queryParam += `&requestedBy=${query.requestedBy}`;
  }
```

MayankBansal12 commented on mock-data/requests/index.js:
```diff
@@ -20,20 +20,18 @@ const pendingRequest = {
-      updatedAt: 1711482912686,
-    },
-  ],
+  data: {
+    id: 'Wl4TTbpSrQDIjs6KLJwD',
+    createdAt: 1711439903761,
+    requestedBy: 'V4rqL1aDecNGoa1IxiCu',
+    from: 1712275200000,
+    until: 1712448000000,
+    type: 'OOO',
+    message: 'request message',
+    lastModifiedBy: 'V4rqL1aDecNGoa1IxiCu',
+    state: 'APPROVED',
+    updatedAt: 1711482912686,
+  },
```
will it not return an array? if not, then why have next and prev cursor?

AnujChhikara replied:
This mock data is used to fetch a specific request that is being approved or rejected. The API ensures that when a superuser approves or rejects a request, we fetch only that particular request and update it accordingly. This is the [API](https://api.realdevsquad.com/requests?dev=true&id=QjLMT6j1jsAdlr8WWa6B) example.

MayankBansal12 replied:
Understood. The mock data has prev and next cursor along with data, i guess that can be removed now!

Achintya-Chatterjee commented on requests/script.js:
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
+
+let debounceTimeout;
+
+userNameFilterInput.addEventListener('input', function () {
+  clearTimeout(debounceTimeout);
+
+  const username = this.value.trim();
+  const suggestionContainer = document.getElementById('userSuggestionsList');
+
+  if (username.length > 2) {
+    debounceTimeout = setTimeout(async () => {
+      const users = await getUsersByUsername(username);
+
+      suggestionContainer.classList.remove('hidden');
+      suggestionContainer.innerHTML = '';
+
+      if (users && users.length > 0) {
+        users.forEach((user) => {
+          const userDiv = document.createElement('div');
+          userDiv.className = 'suggestion';
+          userDiv.textContent = user.username;
+
+          suggestionContainer.appendChild(userDiv);
+
+          userDiv.addEventListener('click', async () => {
+            userNameFilterInput.value = user.username;
```
We should optimize debounce logic to prevent redundant API calls.
 - The current implementation clears the timeout but still allows multiple requests if the user types quickly.  
 - This can cause unnecessary load on the server and degrade performance.  
- Consider using _.debounce() from lodash or write a custom debounce function with a trailing execution strategy.  
- This ensures the API call is made only after the user stops typing, improving efficiency.  
- Additionally, caching previous results for the same input can further reduce API requests.

Achintya-Chatterjee replied:
we should write a custom debounce function, I think this will be the most sought-after approach.

AnujChhikara replied:
Okay, we already have a custom function for debounce , I have used that one.

Achintya-Chatterjee replied:
Cool

Achintya-Chatterjee commented on requests/script.js:
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
+      }
+    }, 300);
+  } else {
+    suggestionContainer.classList.add('hidden');
+    suggestionContainer.innerHTML = '';
```
Clearing and repopulating `suggestionContainer (suggestionContainer.innerHTML = '';`) is an **anti-pattern**. Instead, We should use `DocumentFragment` to append elements in one go.
Also, if users re-type the same input, you're re-fetching data instead of caching previous results

Achintya-Chatterjee commented on requests/util.js:
```diff
@@ -43,6 +43,9 @@ function getQueryParamsString(requestType, query) {
   ) {
     queryParam += `&state=${query.state}`;
   }
+  if (query?.requestedBy != null) {
+    queryParam += `&requestedBy=${query.requestedBy}`;
+  }
```
The query parameters are concatenated directly (`queryParam += '&requestedBy=' + query.requestedBy;`). If `queryParam` is empty, the query string starts with `&`instead of `?`. So we should use `new URLSearchParams(query).toString()` instead of manually appending query parameters.

Achintya-Chatterjee commented on requests/script.js:
```diff
@@ -540,30 +552,35 @@ async function performAcceptRejectAction(isAccepted, e) {
   const parentDiv = e.target.closest('.request__card');
-  parentDiv.classList.add('disabled');
-  const removeSpinner = addSpinner(parentDiv);
+  requestPageContainer.classList.add('disabled');
+  const removeSpinner = addSpinner(requestPageContainer);
```
You are disabling the entire `requestPageContainer (requestPageContainer.classList.add('disabled')`) instead of only disabling the specific card. So this means that one action blocks all other interactions, which is a poor UX decision. Instead, we should only disable the clicked request card.

AnujChhikara replied:
Improved this, now only that card will get disabled when we perform action on that request card

Achintya-Chatterjee commented on requests/util.js:
```diff
@@ -151,3 +154,26 @@ function addRadioButton(labelText, value, groupName) {
+  });
+}
+
+async function getUsersByUsername(username) {
+  try {
+    const res = await fetch(`${API_BASE_URL}/users?search=${username}&size=5`, {
+      credentials: 'include',
+      method: 'GET',
+      headers: {
+        'Content-type': 'application/json',
+      },
+    });
+    const data = await res.json();
+    return data.users;
+  } catch (error) {
+    console.log(error);
+  }
+}
```
If the server returns a non-200 response, you are not checking `res.ok`, which may lead to unexpected behaviours.

AnujChhikara replied:
added that check

Achintya-Chatterjee commented on requests/script.js:
```diff
@@ -105,9 +111,15 @@ function updateUrlWithQuery(type) {
-  let finalUrl =
-    API_BASE_URL +
-    (nextLink || '/requests' + getQueryParamsString(requestType, query));
+  let finalUrl;
+  if (query?.state?.[0] || query?.requestedBy) {
+    finalUrl =
+      API_BASE_URL + '/requests' + getQueryParamsString(requestType, query);
+  } else {
+    finalUrl =
+      API_BASE_URL +
+      (nextLink || '/requests' + getQueryParamsString(requestType, query));
+  }
```
`getRequests()` function has redundant conditions. `if (query?.state?.[0] || query?.requestedBy) {...}` should be refactored. Why are we doing repetitive URL assignments, this is leading to unnecessary complexity.

AnujChhikara replied:
Refactored this function

Achintya-Chatterjee commented on requests/script.js:
```diff
@@ -626,7 +643,92 @@ filterButton.addEventListener('click', (event) => {
 });
 
-applyFilterButton.addEventListener('click', closeFilter);
+applyFilterButton.addEventListener('click', async (event) => {
+  closeFilter();
+
+  const requestData = {
+    state: statusValue,
+    sort: sortByValue,
+  };
+
+  const selectedState = getCheckedValues();
+  const username = userNameFilterInput.value;
+  if (username && username.trim() !== '') {
+    requestData.requestedBy = username;
+  }
+  if (!selectedState) {
+    requestContainer.innerHTML = '';
+    await renderRequestCards(requestData);
+  } else {
+    requestContainer.innerHTML = '';
+    requestData.state = selectedState;
+    await renderRequestCards(requestData);
+  }
+});
```
The function calls `renderRequestCards()` twice, once for `requestData.state === null` and once for `state !== null`. Instead of doing this, we should structure the logic cleanly.

AnujChhikara commented on requests/script.js:
```diff
@@ -626,7 +641,94 @@ filterButton.addEventListener('click', (event) => {
+
+  if (username.length > 2) {
+    debounceTimeout = setTimeout(async () => {
+      const users = await getUsersByUsername(username);
+
+      suggestionContainer.classList.remove('hidden');
+      while (suggestionContainer.firstChild) {
+        suggestionContainer.removeChild(suggestionContainer.firstChild);
+      }
+
+      if (users && users.length > 0) {
+        const fragment = document.createDocumentFragment();
+
+        users.forEach((user) => {
+          const userDiv = document.createElement('div');
+          userDiv.className = 'suggestion';
+          userDiv.textContent = user.username;
+
+          userDiv.addEventListener('click', async () => {
+            userNameFilterInput.value = user.username;
+            suggestionContainer.classList.add('hidden');
+            requestContainer.innerHTML = '';
+
+            const requestData = { requestedBy: user.username };
+            await renderRequestCards(requestData);
+          });
+
+          fragment.appendChild(userDiv);
+        });
+
+        suggestionContainer.appendChild(fragment);
+      }
+    }, 300);
+  } else {
+    suggestionContainer.classList.add('hidden');
+    while (suggestionContainer.firstChild) {
+      suggestionContainer.removeChild(suggestionContainer.firstChild);
+    }
+  }
+});
```
The getUsersByUsername function already has error handling that logs issues and returns an empty array on failure, so the app won’t crash.

AnujChhikara commented:
Hi @Achintya-Chatterjee, thanks for the review! I’ve addressed most of your comments. Regarding your suggestion to cache API calls, we currently don’t have a caching mechanism on the dashboard site. There are several options—like using a Map, a custom function, or other approaches—but I’d love your input on how to choose the best one for our use case. How do you think we should decide?

Achintya-Chatterjee commented on __tests__/requests/requests.test.js:
```diff
@@ -370,5 +381,51 @@ describe('Tests the request cards', () => {
+      }
+    });
+
+    it('should show user suggestions after debounced input', async () => {
```
the test correctly verifies that suggestions appear and include the expected username, it would be beneficial to add assertions for edge cases such as empty results, special characters, or minimum character threshold.

AnujChhikara replied:
added please check
